libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -1401,36 +1401,55 @@ string ABIFunctions::copyToMemoryFunction(bool _fromCalldata)
 
 string ABIFunctions::shiftLeftFunction(size_t _numBits)
 {
+	solAssert(_numBits < 256, "");
+
 	string functionName = "shift_left_" + to_string(_numBits);
 	return createFunction(functionName, [&]() {
-		solAssert(_numBits < 256, "");
-		return
-			Whiskers(R"(
-			function <functionName>(value) -> newValue {
-				newValue := mul(value, <multiplier>)
-			}
-			)")
-			("functionName", functionName)
-			("multiplier", toCompactHexWithPrefix(u256(1) << _numBits))
-			.render();
+		Whiskers templ(R"(
+		function <functionName>(value) -> newValue {
+			newValue := <shiftOp>(value, <secondArg>)
+		}
+		)");
+		templ("functionName", functionName);
+		if (m_evmVersion.hasBitwiseShifting())
+		{
+			templ("shiftOp", "shl");
+			templ("secondArg", to_string(_numBits));
+		}
+		else
+		{
+			templ("shiftOp", "mul");
+			templ("secondArg", toCompactHexWithPrefix(u256(1) << _numBits));
+
+		}
+		return templ.render();
 	});
 }
 
 string ABIFunctions::shiftRightFunction(size_t _numBits, bool _signed)
 {
+	solAssert(_numBits < 256, "");
+
 	string functionName = "shift_right_" + to_string(_numBits) + (_signed ? "_signed" : "_unsigned");
 	return createFunction(functionName, [&]() {
-		solAssert(_numBits < 256, "");
-		return
-			Whiskers(R"(
-			function <functionName>(value) -> newValue {
-				newValue := <div>(value, <multiplier>)
-			}
-			)")
-			("functionName", functionName)
-			("div", _signed ? "sdiv" : "div")
-			("multiplier", toCompactHexWithPrefix(u256(1) << _numBits))
-			.render();
+		Whiskers templ(R"(
+		function <functionName>(value) -> newValue {
+			newValue := <shiftOp>(value, <secondArg>)
```

Comment:
Actually I think this is broken because the top of the stack is `secondArg` for shifts, which in inline assembly would be `shl(<secondArg>, value)`